source/common/grpc/google_async_client_impl.cc

```diff
@@ -0,0 +1,365 @@
+        return Http::HeaderMap::Iterate::Continue;
+      },
+      &ctxt_);
+  // Invoke stub call.
+  rw_ = parent_.stub_->Call(
+      &ctxt_, "/" + service_method_.service()->full_name() + "/" + service_method_.name(), &cq_,
+      tag(Operation::Init));
+  if (rw_ == nullptr) {
+    notifyRemoteClose(Status::GrpcStatus::Unavailable, nullptr, EMPTY_STRING);
+    call_failed_ = true;
+  }
+}
+
+void GoogleAsyncStreamImpl::notifyRemoteClose(Status::GrpcStatus grpc_status,
+                                              Http::HeaderMapPtr trailing_metadata,
+                                              const std::string& message) {
+  if (grpc_status > Status::GrpcStatus::MaximumValid) {
+    grpc_status = Status::GrpcStatus::Unknown;
+  }
+  ENVOY_LOG(debug, "notifyRemoteClose {} {}", grpc_status, message);
+  parent_.stats_.streams_closed_[grpc_status]->inc();
+  callbacks_.onReceiveTrailingMetadata(trailing_metadata ? std::move(trailing_metadata)
+                                                         : std::make_unique<Http::HeaderMapImpl>());
+  callbacks_.onRemoteClose(grpc_status, message);
+}
+
+void GoogleAsyncStreamImpl::sendMessage(const Protobuf::Message& request, bool end_stream) {
+  write_pending_queue_.emplace(request, end_stream);
+  ENVOY_LOG(trace, "Queued message to write ({} bytes)",
+            write_pending_queue_.back().buf_.value().Length());
+  writeQueued();
+}
+
+void GoogleAsyncStreamImpl::closeStream() {
+  // Empty EOS write queued.
+  write_pending_queue_.emplace();
+  writeQueued();
+}
+
+void GoogleAsyncStreamImpl::resetStream() { cleanup(); }
+
+void GoogleAsyncStreamImpl::writeQueued() {
+  if (!call_initialized_ || finish_pending_ || write_pending_ || write_pending_queue_.empty()) {
+    return;
+  }
+  write_pending_ = true;
+  const PendingMessage& msg = write_pending_queue_.front();
+
+  if (!msg.buf_.valid()) {
+    ASSERT(msg.end_stream_);
+    rw_->WritesDone(tag(Operation::WriteLast));
+  } else if (msg.end_stream_) {
+    grpc::WriteOptions write_options;
+    rw_->WriteLast(msg.buf_.value(), write_options, tag(Operation::WriteLast));
+  } else {
+    rw_->Write(msg.buf_.value(), tag(Operation::Write));
+  }
+}
+
+void GoogleAsyncStreamImpl::handleOpCompletion(Operation op, bool ok) {
+  ENVOY_LOG(trace, "handleOpCompletion {} {}", op, ok);
+  // Ignore op completions while CQ is shutting down.
+  if (cq_shutdown_in_progress_) {
+    return;
+  }
+  // Consider failure cases first.
+  if (!ok) {
+    // Early fails can be just treated as Internal.
+    if (op == Operation::Init || op == Operation::ReadInitialMetadata) {
+      notifyRemoteClose(Status::GrpcStatus::Internal, nullptr, EMPTY_STRING);
+      resetStream();
+      return;
+    }
+    // Remote server has closed, we can pick up some meaningful status.
+    if (op == Operation::Read) {
+      finish_pending_ = true;
+      rw_->Finish(&status_, tag(Operation::Finish));
+    }
+    return;
```

Comment:
My assumption here is that a failed `Write[Last]` operation will result in the stream being ended, with the pending `Read` (there is always one pending) then returning false and allowing us to detect this situation. This is what I've seen in various tests for failure. @vjpai to confirm, I will add a TODO for now.

Reply:
What you said sounds good; a client-side Write failure should mean a stream failure/cancellation and eventually a Read will pick up on that if you always have a Read in flight. I suppose that if multiple to-be-Read messages have already been buffered for you at the client side it might take some time, so maybe there's an advantage in processing the Write/Writelast failure early if you don't care about the content of those buffered-up Reads.

Reply:
Thanks for confirming. The current `Grpc::AsyncClient` interface we have inside Envoy that this is exporting doesn't make any guarantee about write failure notification; basically when "bad stuff happens", you get an `onRemoteClose(..)` callback at some point in the future. So, to keep this simple for now we'll just pick up on the `Read` failure.